crates/milli/src/update/facet/new_incremental.rs

```diff
@@ -0,0 +1,477 @@
+                )?
+                .next();
+
+            match parent {
+                Some(Ok((parent_key, _parent_value))) => {
+                    // found parent, cache it for next keys
+                    last_parent = Some(parent_key.left_bound.to_owned());
+
+                    // add to modified list for parent level
+                    touched_parents.push(parent_key.left_bound.to_owned());
+                    self.compute_parent_group(wtxn, child_level, child)?;
+                }
+                Some(Err(err)) => return Err(err.into()),
+                None => {
+                    self.compute_parent_group(wtxn, child_level, child)?;
+                    break;
+                }
+            }
+        }
+        // do we have children without parents?
+        if let Some(child) = touched_children.next() {
+            // no parent for that key
+            let mut it = self
+                .db
+                .remap_data_type::<DecodeIgnore>()
+                .prefix_iter_mut(wtxn, &parent_level_left_bound)?;
+            match it.next() {
+                // 1. left of the current left bound, or
+                Some(Ok((first_key, _first_value))) => 'change_left_bound: {
+                    // make sure we don't spill on the neighboring fid (level also included defensively)
+                    if first_key.field_id != self.field_id || first_key.level != parent_level {
+                        break 'change_left_bound;
+                    }
+                    // remove old left bound
+                    unsafe { it.del_current()? };
+                    drop(it);
+                    // pop all elements and order to visit the new left bound
+                    touched_parents.push(child.clone());
+                    self.compute_parent_group(wtxn, child_level, child)?;
+                    for child in touched_children {
+                        let new_left_bound = touched_parents.last_mut().unwrap();
+                        new_left_bound.clear();
+                        new_left_bound.extend_from_slice(&child);
+                        self.compute_parent_group(wtxn, child_level, child)?;
+                    }
+                }
+                Some(Err(err)) => return Err(err.into()),
+                // 2. max level reached, exit
+                None => {
+                    drop(it);
+                    self.compute_parent_group(wtxn, child_level, child)?;
+                    for child in touched_children {
+                        self.compute_parent_group(wtxn, child_level, child)?;
+                    }
+                }
+            }
+        }
+        if !touched_parents.is_empty() {
+            self.find_touched_parents(
+                wtxn,
+                parent_level,
+                touched_parents
+                    // no need to `rev` here because the parents were already visited in reverse order
+                    .into_iter(),
+            )?;
+        }
```

Comment:
I understand the recursive here, I am completely ok with it. I just want your thoughts about doing an iterative implementation using a BTreeSet<(level, Vec<u8>)>?

Maybe poping and pushing in a BTreeSet is not the best idea, and the recursive calls can't be that high. 🤔

Reply:
the recursive calls are max 255, since level is a u8 and we exit early if it is u8::MAX. Also to reach 255 level we need a number of nodes whose `log4` equals 255, which requires approximatively `1_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000` nodes. Actual values we reach is level 12 which requires 15 millions of nodes already. `u32::MAX` (max number of documents) fits in 16 levels.

However, as the call is terminal, it is true that it could be rewritten as a for loop, no BTree necessary.